src/visitors/hoist.js

```diff
@@ -183,122 +182,201 @@ module.exports = {
+        );
 
-            return getNamedExportVarDecl(
+        path.insertAfter(
+          EXPORT_ASSIGN_TEMPLATE({
+            EXPORTS: getExportsIdentifier(asset),
+            NAME: t.identifier(specifier.exported.name),
+            LOCAL: getIdentifier(
               asset,
-              specifier.exported,
-              specifier.local
-            );
+              'import',
+              source.value,
+              specifier.local.name
+            )
           })
         );
       }
 
-      if (declarations.length) {
-        path.replaceWith(t.variableDeclaration('var', declarations));
+      path.remove();
+    } else if (declaration) {
+      path.replaceWith(declaration);
+
+      if (
+        t.isFunctionDeclaration(declaration) ||
+        t.isClassDeclaration(declaration)
+      ) {
+        addExport(asset, path, declaration.id, declaration.id);
+      } else if (t.isVariableDeclaration(declaration)) {
+        for (let d of declaration.declarations) {
+          if (t.isIdentifier(d.id)) {
+            addExport(asset, path, d.id, d.id);
+          } else if (t.isObjectPattern(d.id)) {
+            for (let prop of d.id.properties) {
+              if (t.isAssignmentPattern(prop.value)) {
+                // TODO
+              } else if (t.isRestProperty(prop)) {
+                // TODO
+              }
+            }
+          } else if (t.isArrayPattern(d.id) && d.id.elements) {
+            // TODO
+          }
+        }
       }
+    } else if (specifiers.length > 0) {
+      for (let specifier of specifiers) {
+        addExport(asset, path, specifier.local, specifier.exported);
+      }
+
+      path.remove();
     }
+
+    // Mark the asset as an ES6 module, so we handle imports correctly in the packager.
+    asset.cacheData.isES6Module = true;
+  },
+
+  ExportAllDeclaration(path, asset) {
+    // if (BREAK_COMMON_JS) {
+    //   path.replaceWith(
+    //     t.identifier(
+    //       '$' +
+    //         asset.id +
+    //         '$expand_exports$' +
+    //         t.toIdentifier(path.node.source.value)
+    //     )
+    //   );
+    // } else {
+    //   path.replaceWith(
+    //     EXPORT_ALL_TEMPLATE({
+    //       EXPORTS: t.identifier('$' + asset.id + '$exports'),
+    //       SOURCE: t.identifier(
+    //         '$' +
+    //           asset.id +
+    //           '$require$' +
+    //           t.toIdentifier(path.node.source.value)
+    //       )
+    //     })
+    //   );
+    // }
+
+    asset.cacheData.isES6Module = true;
   }
 };
 
-function getNamedExportVarDecl(asset, name, init) {
-  let varName = getNamedExportIdentifierName(asset, name);
+function addExport(asset, path, local, exported) {
+  asset.cacheData.exports[getName(asset, 'export', exported.name)] =
+    exported.name;
 
-  return NAMED_EXPORT_TEMPLATE({
-    NAME: t.identifier(varName),
-    EXPORTS: getExportsIdentifier(asset),
-    BINDING: name,
-    INIT: init
-  }).declarations[0];
+  path.insertAfter(
```

Comment:
For each export, we don't create a new variable, we just rename the local variable within the module to the exported name. Also, we add a property on the exports object for use in namespaces and commonjs imports of ES6 modules.